File: iOS/Kaare/Kaare.m
#import "Kaare.h"

NSString* const KaareErrorDomain = @"com.kaare.KaareErrorDomain";

@implementation Kaare
{
    NSMutableDictionary* _localCommands;
    id<KaareTransport> _transport;
}

-(id<KaareTransport>)transport { return _transport; }

-(instancetype)initWithTransport:(id<KaareTransport>)transport
{
    if (self = [super init])
    {
        _localCommands = [@{} mutableCopy];
        _transport = transport;
        [_transport onIncomingCommand:^(NSString *cmd, NSArray *params) {
            if (_localCommands[cmd])
            {
                CommandHandler handler = _localCommands[cmd];
                return handler(params);
            }
            
            return [RACSignal error:[NSError errorWithDomain:KaareErrorDomain
                                                        code:KaareErrCommandNotFound
                                                    userInfo:@{NSLocalizedDescriptionKey:[NSString stringWithFormat:@"Command %@ cannot be found",cmd]}]];
        }];
    }
    return self;
}

-(RACSignal*)executeCommand:(NSString*)cmd params:(NSArray*)params
{
    return [_transport executeCommand:cmd params:params];
}

-(void)registerCommand:(NSString *)cmd handler:(CommandHandler)handler
{
    if (_localCommands[cmd])
        [NSException raise:@"KaareDuplicateCommand" format:@"Kaare already contains a command with name %@. Consider using different name",cmd];
    
    _localCommands[cmd] = handler;
}


@end

File: js/src/kaare.js
var Kaare = function (transport = new Kaare.transports.Native()) { // jshint ignore:line 
    let _localCommands = {},
        _transport = transport

    _transport.onIncomingCommand = (cmd, params) => {
        if (cmd in _localCommands)
        {
            let handler = _localCommands[cmd],
                handlerResult
            
            try { handlerResult = handler(params) }
            catch (ex) { return Rx.Observable.throw(ex) }

            return handlerResult
        }
        return Rx.Observable.throw(new Error(`Command ${cmd} cannot be found`))
    }

    this.executeCommand = (cmd, params) => (_transport.executeCommand(cmd, params))

    this.registerCommand = (cmd, handler) => {
        if (cmd in _localCommands)
            throw new Error(`Kaare already contains a command with name ${cmd}. Consider using different name`)

        _localCommands[cmd] = handler
    }

    this.transport = _transport
}

Kaare.transports = {}
